src/Components/Analyzers/src/ComponentParameterUsageAnalyzer.cs

```diff
@@ -0,0 +1,105 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Operations;
+
+namespace Microsoft.AspNetCore.Components.Analyzers
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    public class ComponentParameterUsageAnalyzer : DiagnosticAnalyzer
+    {
+        public ComponentParameterUsageAnalyzer()
+        {
+            SupportedDiagnostics = ImmutableArray.Create(new[]
+            {
+                DiagnosticDescriptors.ComponentParametersShouldNotBeSetOutsideOfTheirDeclaredComponent,
+            });
+        }
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; }
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.Analyze | GeneratedCodeAnalysisFlags.ReportDiagnostics);
+            context.RegisterCompilationStartAction(context =>
+            {
+                if (!ComponentSymbols.TryCreate(context.Compilation, out var symbols))
+                {
+                    // Types we need are not defined.
+                    return;
+                }
+
+                context.RegisterOperationBlockStartAction(startBlockContext =>
+                {
+                    startBlockContext.RegisterOperationAction(context =>
+                    {
+                        var assignmentOperation = (IAssignmentOperation)context.Operation;
+                        var leftHandSide = assignmentOperation.Target;
+                        if (leftHandSide == null)
+                        {
+                            // Malformed assignment, no left hand side.
+                            return;
+                        }
+
+                        if (leftHandSide.Kind != OperationKind.PropertyReference)
+                        {
+                            // We don't want to capture situations where a user does
+                            // MyOtherProperty = aComponent.SomeParameter
+                            return;
+                        }
+
+                        var propertyReference = (IPropertyReferenceOperation)leftHandSide;
+                        var componentProperty = propertyReference.Member;
```

Comment:
So, this is definitely `IPropertySymbol` so you don't need `HasParameterAttribute`.